File: lru_cache_DLL.py
# Key Idea: since we need to perform remove and add operations, using singly linked list will not be efficient. For removal, take time O(n)
# create node class to store doubly linked list

class Node:
    def __init__(self, key=-1, val=-1):
        self.next = None
        self.prev = None
        self.key = key
        self.val = val
       
class LRUCache:

    def __init__(self, capacity: int):
        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        # to store key as given keys and value as pointer to corresponding doubly linked list node
        self.hashmap = {}
        # to create doubly linked list
        self.head.next = self.tail
        self.tail.prev = self.head
        self.capacity = capacity
    
    def addToHead(self, node) -> None:
        #TC: O(1)
        node.next = self.head.next
        node.prev = self.head
        self.head.next = node
        node.next.prev = node
    
    def removeNode(self, node) -> None:
        #TC : O(1)
        node.next.prev = node.prev
        node.prev.next = node.next

    def get(self, key: int) -> int:
        #TC: O(1)
        if not key in self.hashmap:
            return -1
        node = self.hashmap[key]
        self.removeNode(node)
        self.addToHead(node)
        return node.val
        

    def put(self, key: int, value: int) -> None:
        #TC:O(1)
        # if key is already present, update the exisiting key value in hashmap, remove node from DLL, add current node in DLL
        if  key in self.hashmap:
            node = self.hashmap[key]
            node.val = value
            self.removeNode(node)
            self.addToHead(node)
        else:
            # if capacity is full, remove both from hashmap and linked list
            if self.capacity == len(self.hashmap):
                tail_prev = self.tail.prev
                self.removeNode(tail_prev)
                del self.hashmap[tail_prev.key]
            
            # if not present, create new node and add to the head and update hashmap
            new_node = Node(key, value)
            self.addToHead(new_node)
            self.hashmap[key] = new_node
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: LRUCacheDLL.java
//Microsoft
////Time complexity = O(1) for get and put
//SC =O(capacity)
//Using Doubly LL could be suitable because it has two way directions to node ,So head pointing to first node of the pointer and then ass a second element ex.PU(2,2) then first element should be pushed to the next node and add recent element after the head.Double LL used to maintain order to remove or add an element .In add method the node.next = head.next,node.prev = head ,node.next = node,node.next.prev = node and in remove method remove the element and add the element because size capacity = 4 is considered in DLL.
class LRUCache {
    class Node{
        int key,value;
        Node prev,next;
        public Node(int key, int value){
            this.key = key;
            this.value = value;
        }
    }
    Node head,tail;
    HashMap<Integer,Node> map;
    int capacity;
    
    private void removeNode(Node node){
        node.next.prev = node.prev;
        node.prev.next = node.next;
    }
    
    private void addToHead(Node node){
        node.next = head.next;
        node.prev = head;
        head.next = node;
        node.next.prev = node;
    }
    
    public LRUCache(int capacity) {
        map = new HashMap<>();
        this.capacity = capacity;
        head = new Node(-1,-1);
        tail = new Node(-1,-1);
        head.next = tail;
        tail.prev = head;
    }
    //Time complexity = O(1)
    public int get(int key) {
        if(!map.containsKey(key)) return -1;
        Node node = map.get(key);
        removeNode(node);
        addToHead(node);
        return node.value;
    }
    //Time complexity = O(1)
    public void put(int key, int value) {
        if(map.containsKey(key)){
            Node node = map.get(key);
            node.value = value;
            removeNode(node);
            addToHead(node);
            return;
        }
        if(map.size() == capacity){
            Node tailPrev = tail.prev;
            removeNode(tailPrev);
            map.remove(tailPrev.key);
        }
        Node node = new Node(key,value);
        addToHead(node);
        map.put(key,node);
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */